bug: mssql pandas data can't be converted by correct type

### What happened?

If I send a query statement to mssql, the result data cannot be converted by paired type.

Because ibis changes the `_get_schema_using_query` of mssql at `9.3.0`
```python
# For some reason when using "Latin1_General_100_BIN2_UTF8"
# the stored procedure `sp_describe_first_result_set` starts throwing errors about DLL loading.
# This "dynamic management function" uses the same algorithm and allows
# us to pre-filter the columns we want back.
# The syntax is:
# `sys.dm_exec_describe_first_result_set(@tsql, @params, @include_browse_information)`
query = f"""SELECT name,
                is_nullable AS nullable,
                system_type_name,
                precision,
                scale
            FROM
                sys.dm_exec_describe_first_result_set({tsql}, NULL, 0)"""
with self._safe_raw_sql(query) as cur:
    rows = cur.fetchall()
```
the rows will get
```
[('orderkey', True, 'int', 10, 0), ('custkey', True, 'int', 10, 0), ('orderstatus', True, 'varchar(max)', 0, 0), ('totalprice', True, 'varchar(max)', 0, 0), ('orderdate', True, 'date', 10, 0) .....]
```
different before
```
[(False, 1, 'orderkey', True, 56, 'int', 4, 10, 0, None, None, None, None, None, None, None, None, None, None, False, False, False, None, None, None, None, None, False, None, True, False, False, None, None, None, 38, 4, None, None), (False, 2, 'custkey', True, 56, 'int', 4, 10, 0, None, None, None, None, None, None, None, None, None, None, False, False, False, None, None, None, None, None, False, None, True, False, False, None, None, None, 38, 4, None, None) ......]
```
But we still use `sorted(rows, key=itemgetter(1))` to iterate rows. It will make the order wrong.
```python
for (
    name,
    nullable,
    system_type_name,
    precision,
    scale,
) in sorted(rows, key=itemgetter(1)):
```

### What version of ibis are you using?

9.3.0

### What backend(s) are you using, if any?

mssql

### Relevant log output

```sh
../app/model/connector.py:19: in query
    return self.connection.sql(sql).limit(limit).to_pandas()
../.venv/lib/python3.11/site-packages/ibis/expr/types/relations.py:3568: in to_pandas
    return self.execute(**kwargs)
../.venv/lib/python3.11/site-packages/ibis/expr/types/core.py:404: in execute
    return self._find_backend(use_default=True).execute(
../.venv/lib/python3.11/site-packages/ibis/backends/sql/__init__.py:289: in execute
    result = self._fetch_from_cursor(cur, schema)
../.venv/lib/python3.11/site-packages/ibis/backends/sql/__init__.py:106: in _fetch_from_cursor
    df = PandasData.convert_table(df, schema)
../.venv/lib/python3.11/site-packages/ibis/formats/pandas.py:119: in convert_table
    columns.append(cls.convert_column(series, dtype))
../.venv/lib/python3.11/site-packages/ibis/formats/pandas.py:151: in convert_column
    result = convert_method(obj, dtype, pandas_type)
../.venv/lib/python3.11/site-packages/ibis/formats/pandas.py:234: in convert_Date
    return s.map(try_date, na_action="ignore")
../.venv/lib/python3.11/site-packages/pandas/core/series.py:4700: in map
    new_values = self._map_values(arg, na_action=na_action)
../.venv/lib/python3.11/site-packages/pandas/core/base.py:921: in _map_values
    return algorithms.map_array(arr, mapper, na_action=na_action, convert=convert)
../.venv/lib/python3.11/site-packages/pandas/core/algorithms.py:1745: in map_array
    return lib.map_infer_mask(
lib.pyx:2895: in pandas._libs.lib.map_infer_mask
    ???
lib.pyx:2932: in pandas._libs.lib._map_infer_mask
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

v = '1_370'

    def try_date(v):
        if isinstance(v, datetime.datetime):
            return v.date()
        elif isinstance(v, str):
            if v.endswith("Z"):
                return datetime.datetime.fromisoformat(v[:-1]).date()
>           return datetime.date.fromisoformat(v)
E           ValueError: Invalid isoformat string: '1_370'

../.venv/lib/python3.11/site-packages/ibis/formats/pandas.py:230: ValueError
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/ibis/backends/mssql/__init__.py
+++ b/ibis/backends/mssql/__init__.py
@@ -244,24 +244,21 @@ class Backend(SQLBackend, CanCreateCatalog, CanCreateDatabase, CanCreateSchema, 
         # us to pre-filter the columns we want back.
         # The syntax is:
         # `sys.dm_exec_describe_first_result_set(@tsql, @params, @include_browse_information)`
-        query = f"""SELECT name,
-                        is_nullable AS nullable,
-                        system_type_name,
-                        precision,
-                        scale
-                    FROM
-                        sys.dm_exec_describe_first_result_set({tsql}, NULL, 0)"""
+        query = f"""
+        SELECT
+          name,
+          is_nullable,
+          system_type_name,
+          precision,
+          scale
+        FROM sys.dm_exec_describe_first_result_set({tsql}, NULL, 0)
+        ORDER BY column_ordinal
+        """
         with self._safe_raw_sql(query) as cur:
             rows = cur.fetchall()
 
         schema = {}
-        for (
-            name,
-            nullable,
-            system_type_name,
-            precision,
-            scale,
-        ) in sorted(rows, key=itemgetter(1)):
+        for name, nullable, system_type_name, precision, scale in rows:
             newtyp = self.compiler.type_mapper.from_string(
                 system_type_name, nullable=nullable
             )

--- a/ibis/backends/mssql/tests/test_client.py
+++ b/ibis/backends/mssql/tests/test_client.py
@@ -81,6 +81,29 @@ def test_get_schema(con, server_type, expected_type, temp_table):
     assert con.sql(f"SELECT * FROM [{temp_table}]").schema() == expected_schema
 
 
+def test_schema_type_order(con, temp_table):
+    columns = []
+    pairs = {}
+
+    for i, db_type in enumerate(DB_TYPES):
+        if isinstance(db_type, tuple) and len(db_type) == 2:
+            server_type, expected_type = db_type
+            column_name = f"col_{i}"
+            columns.append(f"{column_name} {server_type}")
+            pairs[column_name] = expected_type
+
+    stmt = f"CREATE TABLE [{temp_table}] ({', '.join(columns)})"
+
+    with con.begin() as c:
+        c.execute(stmt)
+
+    expected_schema = ibis.schema(pairs)
+
+    assert con.get_schema(temp_table) == expected_schema
+    assert con.table(temp_table).schema() == expected_schema
+    assert con.sql(f"SELECT * FROM [{temp_table}]").schema() == expected_schema
+
+
 def test_builtin_scalar_udf(con):
     @udf.scalar.builtin
     def difference(a: str, b: str) -> int:

--- a/ibis/backends/mssql/tests/test_client.py
+++ b/ibis/backends/mssql/tests/test_client.py
@@ -1,13 +1,15 @@
 from __future__ import annotations
 
 import pytest
+import sqlglot as sg
+import sqlglot.expressions as sge
 from pytest import param
 
 import ibis
 import ibis.expr.datatypes as dt
 from ibis import udf
 
-DB_TYPES = [
+RAW_DB_TYPES = [
     # Exact numbers
     ("BIGINT", dt.int64),
     ("BIT", dt.boolean),
@@ -36,23 +38,9 @@ DB_TYPES = [
     ("DATETIME", dt.Timestamp(scale=3)),
     # Characters strings
     ("CHAR", dt.string),
-    param(
-        "TEXT",
-        dt.string,
-        marks=pytest.mark.notyet(
-            ["mssql"], reason="Not supported by UTF-8 aware collations"
-        ),
-    ),
     ("VARCHAR", dt.string),
     # Unicode character strings
     ("NCHAR", dt.string),
-    param(
-        "NTEXT",
-        dt.string,
-        marks=pytest.mark.notyet(
-            ["mssql"], reason="Not supported by UTF-8 aware collations"
-        ),
-    ),
     ("NVARCHAR", dt.string),
     # Binary strings
     ("BINARY", dt.binary),
@@ -67,6 +55,23 @@ DB_TYPES = [
     ("GEOGRAPHY", dt.geography),
     ("HIERARCHYID", dt.string),
 ]
+PARAM_TYPES = [
+    param(
+        "TEXT",
+        dt.string,
+        marks=pytest.mark.notyet(
+            ["mssql"], reason="Not supported by UTF-8 aware collations"
+        ),
+    ),
+    param(
+        "NTEXT",
+        dt.string,
+        marks=pytest.mark.notyet(
+            ["mssql"], reason="Not supported by UTF-8 aware collations"
+        ),
+    ),
+]
+DB_TYPES = RAW_DB_TYPES + PARAM_TYPES
 
 
 @pytest.mark.parametrize(("server_type", "expected_type"), DB_TYPES, ids=str)
@@ -85,14 +90,23 @@ def test_schema_type_order(con, temp_table):
     columns = []
     pairs = {}
 
-    for i, db_type in enumerate(DB_TYPES):
-        if isinstance(db_type, tuple) and len(db_type) == 2:
-            server_type, expected_type = db_type
-            column_name = f"col_{i}"
-            columns.append(f"{column_name} {server_type}")
-            pairs[column_name] = expected_type
+    quoted = con.compiler.quoted
+    dialect = con.dialect
+    table_id = sg.to_identifier(temp_table, quoted=quoted)
 
-    stmt = f"CREATE TABLE [{temp_table}] ({', '.join(columns)})"
+    for i, (server_type, expected_type) in enumerate(RAW_DB_TYPES):
+        column_name = f"col_{i}"
+        columns.append(
+            sge.ColumnDef(
+                this=sg.to_identifier(column_name, quoted=quoted), kind=server_type
+            )
+        )
+        pairs[column_name] = expected_type
+
+    query = sge.Create(
+        kind="TABLE", this=sge.Schema(this=table_id, expressions=columns)
+    )
+    stmt = query.sql(dialect)
 
     with con.begin() as c:
         c.execute(stmt)
@@ -101,7 +115,9 @@ def test_schema_type_order(con, temp_table):
 
     assert con.get_schema(temp_table) == expected_schema
     assert con.table(temp_table).schema() == expected_schema
-    assert con.sql(f"SELECT * FROM [{temp_table}]").schema() == expected_schema
+
+    raw_sql = sg.select("*").from_(table_id).sql(dialect)
+    assert con.sql(raw_sql).schema() == expected_schema
 
 
 def test_builtin_scalar_udf(con):

